### DISCUSSION TITLE
how to repplace  the afl++ havoc strategy(mopt) with other strategy ?

### DISCUSSION THREAD

**Discussion by ricardo-kcc** [AFLplusplus/AFLplusplus#1477] [General]
I try to  repplace  the afl++ havoc strategy(mopt) with other strategy. But I don't know how to do. the custom mutator API seems not to implent it. Could anyone help me ?

**Comment by vanhauser-thc** [comment]
you find the normal havoc and the mopt implementation in src/afl-fuzz-one.c
you could also implement your own strategy (if you mean not only mutation but learning what mutation was successful) via the custom mutators (e.g. learning that an afl_custom_fuzz() mutation resulted in new coverage can be picked up via afl_custom_queue_new_entry()).

**Comment by ricardo-kcc** [comment]
hi, @vanhauser-thc . I want to use multi-armed bandit model for the random havoc stage. It need to modify the fuzz_one() function, so I should copy the modified fuzz_one() into afl_custom_fuzz() ?

**Comment by vanhauser-thc** [comment]
no it is much more complicated than that. for your purposes it would be easier to fork afl++ and modify afl-fuzz-one.c instead.
as a custom mutator you would need to copy a lot of code form afl-fuzz-one.c to the afl_custom_fuzz function, and mutate depending on a state you keep and what the bandit says, and pick up on new coverage via that queue_new_entry function to update the data for the bandit.
if you dont know what you are doing that is more work (and would be slower) than modifying afl++ itself.

**Reply by ricardo-kcc** [reply]
I wnat to implement like [this](https://github.com/MagicHavoc/Havoc-Study/tree/main/fuzzers/Havoc_DMA), but I found it is hard to directly modify the afl++' afl-fuzz-one.c  because it have original fuzz_one and mopt mode.

**Comment by domenukk** [comment]
Maybe take a look at LibAFL, it should allow you to do this without forking
